detection_rules/rule.py

```diff
@@ -300,10 +360,25 @@ def setup_in_note(self, value: bool):
     def skip_validate_note(self) -> bool:
         return os.environ.get('DR_BYPASS_NOTE_VALIDATION_AND_PARSE') is not None
 
+    def _validate_entries(self, node: MarkoDocument):
```

Comment:
This feels brittle, based on consistency by rule authors to organize a free form field correctly every time.

Rather than going from `parse note -> pattern match objects -> validate`, how do you feel about the alternative [approach](https://github.com/elastic/detection-rules/issues/2584#issuecomment-1441026155) of `new fields -> validate -> insertion in to_api_format`.

This is a lot less sensitive to authorship inconsistencies

Reply:
looks like the options would be to either do it:
1. in the `to_dict` method within the mixin. This would inherit in the `to_api_format` method
2. solely in the `to_api_format` method, acting on the post dumped dict

1. seems like a better option, but _could_ be confusing


https://github.com/elastic/detection-rules/blob/c3d8bac40281b496fb4a1a717390b7626c03effe/detection_rules/rule.py#L1011-L1019

https://github.com/elastic/detection-rules/blob/c3d8bac40281b496fb4a1a717390b7626c03effe/detection_rules/mixins.py#L111-L119

Reply:
looks like certain line breaks are legal

![image](https://user-images.githubusercontent.com/16747370/220983029-fa3024f6-ab6e-4160-af97-3a28a411e142.png)

![image](https://user-images.githubusercontent.com/16747370/220983094-f5b21913-8c4f-4edf-a530-b6adf340f77a.png)

Reply:
> Rather than going from `parse note -> pattern match objects -> validate`, how do you feel about the alternative [approach](https://github.com/elastic/detection-rules/issues/2584#issuecomment-1441026155) of `new fields -> validate -> insertion in to_api_format`.


After internal discussion with @terrancedejesus and @eric-forte-elastic here's the major takeaways.

- BLUF: Redesigning/reenginnering how we insert osquery and insight entries is out of scope for the initial request.
- We're left with two options:
  1. Merge this PR as-is, and open a new issue to track the work to be done to improve the osquery/insight process. This at least will cover the osquery entries already added in the note field to rules in the interim.
  2. Close this PR (don't merge) or hold and redesign the issue/PR, until it includes all of the new changes. (Note: the insights need to be in by 8.7)

We've said for a while that adding new features to the note field is not a good idea, and we should instead add new fields to the rule schema (even if we strip them out). This provides more seperation, flexibility, and better future proofs the features in terms of validation, versioning, and maintenance.
- Migrating the osquery and insight entries to the rule schema will require the following changes.
  - Adding insights and osquery fields to the rule schema so that they can be validated.
    -  E.x.
    ```toml
      [rule.insights]
      windows_process = '{"label":"Investigate activities done by the subject process", "providers": [[{"field": "process.entity_id", "value": "process.entity_id", "type":  "parameter"}]]}'
      windows_host = '{"label":"Investigate activities done by the subject host", "providers": [[{"field": "host.name", "value": "host.name", "type":  "parameter"}]]}'

      [rule.osqueries]
      windows_dns = '{"query":"SELECT * FROM dns_cache", "label":"Osquery - Retrieve DNS Cache"}'

    ```
  - Note: This format in the toml will enable single or multiline entries.
  - Note: We can validate each entry defined by dataclass schema
- Strip out `rule.insights` and `rule.osqueries`  (e.g. `rule.to_api_format()`)
- Add a unit test or add to validation to make sure each entry specified in the rule toml is used at least once in the `note` field.
- Refactor all rules with existing osquery and insight entries to use the new format.

Reply:
I think we can just update it here.

Also, we would expand out the object, since we would have a schema as code for validation anyway. Makes it less error prone for authors.

Reply:
We could render them at once like this:

```python

>>> my_insight = "something else"
>>> my_osquery = "something here"
>>> note = "{my_osquery}\n\n{my_insight}"
>>> note.format(**locals())
'something here\n\nsomething else'
```

But will still need the unit test to check to make sure the entry is used at least once.